numba/tests/__init__.py

```diff
@@ -82,22 +92,59 @@ def _getParentArgParser(self):
         parser.add_argument('-m', '--multiprocess', dest='multiprocess',
                             action='store_true',
                             help='Parallelize tests')
+        parser.add_argument('-l', '--list', dest='list',
+                            action='store_true',
+                            help='List tests without running them')
         parser.add_argument('--profile', dest='profile',
                             action='store_true',
                             help='Profile the test run')
         return parser
 
     def parseArgs(self, argv):
+        if '-l' in argv:
+            argv.remove('-l')
+            self.list = True
         if PYVERSION < (3, 4) and '-m' in argv:
             # We want '-m' to work on all versions, emulate this option.
             argv.remove('-m')
             self.multiprocess = True
         super(NumbaTestProgram, self).parseArgs(argv)
+        # in Python 2.7, the 'discover' option isn't implicit.
+        # So if no test names were provided and 'self.test' is empty,
+        # we assume discovery hasn't been done yet.
+        if (not getattr(self, 'testNames', None) and
+            not self.test or (isinstance(self.test, suite.BaseTestSuite) and
+                              not self.test.countTestCases())):
+            self._do_discovery([])
+            
         if self.verbosity <= 0:
             # We aren't interested in informational messages / warnings when
             # running with '-q'.
             self.buffer = True
-
+            
+    def _do_discovery(self, argv, Loader=None):
```

Comment:
They aren't, but in fact, they should be (for completeness). 
With Python 2.x there was an explicit 'discover' argument, which for Python 3.x seems implicit.
With my current patch the `python -m numba.testing discover a b c` would ignore 'a', 'b' and 'c', but would work fine if spelled `python -m numba.testing a b c`.

The unittest code (and logic) is quite messy, with many different ways to solve the same problem. I'm still reverse-engineering bits and pieces to make sure I'm not altering or removing functionality in my patches.